components/web_discovery/browser/content_scraper.h

```diff
@@ -0,0 +1,116 @@
+  PageScrapeResult& operator=(const PageScrapeResult&) = delete;
+
+  base::Value SerializeToValue();
+  static std::unique_ptr<PageScrapeResult> FromValue(const base::Value& dict);
+
+  GURL url;
+  // A map of DOM selectors to list of scraped values embedded in a Dict.
+  // Each dict contains arbitrary keys (defined in the patterns) to scraped
+  // values.
+  base::flat_map<std::string, std::vector<base::Value::Dict>> fields;
+  std::string id;
+
+  // Only available for non-strict scrapes with "searchQuery"/"widgetTitle"
+  // scrape rules
+  std::optional<std::string> query;
+};
+
+// Extracts attribute values from the page DOM for reporting purposes.
+// ContentScraper utilizes the following techniques:
+//
+// a) Extraction within the current page in the renderer (via `ScrapePage`).
+//    The `mojom::DocumentExtractor` is used to request attribute values
+//    from the current DOM in the view. Typically, this is used to exact a
+//    search query, and decide whether the page is worthy of investigation
+//    and reporting.
+// b) Parsing and extracting HTML from a double fetch. This follows
+//    the extraction in a). Used to extract all other needed details
+//    from the page i.e. search results. Uses a Rust library for DOM
+//    operations, in respect of Rule of Two.
+class ContentScraper {
+ public:
+  using PageScrapeResultCallback =
+      base::OnceCallback<void(std::unique_ptr<PageScrapeResult>)>;
+
+  explicit ContentScraper(const ServerConfigLoader* server_config_loader);
+  ~ContentScraper();
+
+  ContentScraper(const ContentScraper&) = delete;
+  ContentScraper& operator=(const ContentScraper&) = delete;
+
+  // For initial page scrape in renderer
+  void ScrapePage(const GURL& url,
+                  bool is_strict_scrape,
+                  mojom::DocumentExtractor* document_extractor,
+                  PageScrapeResultCallback callback);
+  // For subsequent double fetches after initial scrape
+  void ParseAndScrapePage(const GURL& url,
+                          bool is_strict_scrape,
+                          std::unique_ptr<PageScrapeResult> prev_result,
+                          std::string html,
+                          PageScrapeResultCallback callback);
+
+ private:
+  void ProcessStandardRule(const std::string& report_key,
+                           const ScrapeRule& rule,
+                           const std::string& root_selector,
+                           const GURL& url,
+                           PageScrapeResult* scrape_result);
+  void OnScrapedElementAttributes(
+      bool is_strict_scrape,
+      std::unique_ptr<PageScrapeResult> scrape_result,
+      PageScrapeResultCallback callback,
+      std::vector<mojom::AttributeResultPtr> attribute_results);
+  void OnRustElementAttributes(
+      bool is_strict_scrape,
+      std::unique_ptr<PageScrapeResult> scrape_result,
+      PageScrapeResultCallback callback,
+      rust::Vec<rust_document_extractor::AttributeResult> attribute_results);
```

Comment:
missing header for cxx.h, but exposing this in the header for ContentScraper is problematic because that exposes it to the web discovery service. Rust types including cxx and lib.rs.h headers should be encpasulated so cxx types are not exposed in the public interface